src/core/ext/filters/client_channel/subchannel.cc

```diff
@@ -503,20 +504,22 @@ class Subchannel::HealthWatcherMap::HealthWatcher
         health_check_service_name_.get(), subchannel_->connected_subchannel_,
         subchannel_->pollset_set_, subchannel_->channelz_node_);
     Ref().release();  // Ref for health callback tracked manually.
-    health_check_client_->NotifyOnHealthChange(&state_, &on_health_changed_);
+    health_check_client_->NotifyOnHealthChange(&pending_state_,
+                                               &on_health_changed_);
   }
 
   static void OnHealthChanged(void* arg, grpc_error* error) {
     auto* self = static_cast<HealthWatcher*>(arg);
     Subchannel* c = self->subchannel_;
     {
       MutexLock lock(&c->mu_);
+      self->state_ = self->pending_state_;
```

Comment:
It looks like this code that reads pending_state_ holds the subchannel lock while the other one that writes the same data is under the health-check-client's lock (SetHealthStatus). As a result, there's no ordering between them and will still have a TSAN issue.

Reply:
I should have just checked the tests, indeed that TSAN race popped up.

Reply:
Actually, I think my last comment was incorrect; I'll drop this review and look at it closely on Monday.

Reply:
Yeah, I don't think this is an issue.  The health check client will update `pending_state_` only before this callback runs, so it's safe to access it here (but nowhere else).  That's why this level of indirection works.